### ISSUE TITLE
Correct incorrect gradients

### ISSUE THREAD

**Issue by Songtech-0912** [elaraproject/elara-math#3]
It has been identified that the library's `matmul()` function frequently gives incorrect gradients. This should be fixed, and tests should be added to the library to ensure that the gradients are the same as those calculated by PyTorch/TensorFlow.
